### PR TITLE
[booster] add tests for ddp and low level zero's checkpointio

### LINKED ISSUE #3716: [booster] update booster tests.

### PR DESCRIPTION
**Pull request by jiangmingyan**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
> 
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

resolved  #3716 

this work is to improve booster unit tests. I add some tests of torchDDPPlugin's and LowLevelZeroPlugin's CheckpointIO.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/torch_ddp_plugin.py [hunk 1]
```diff
@@ -115,6 +116,9 @@ def control_device(self) -> bool:
     def supported_devices(self) -> List[str]:
         return ['cuda']
 
+    def no_sync(self, model) -> contextmanager:
```
Comment (ver217) [inline colossalai/booster/plugin/torch_ddp_plugin.py:119]: `no_sync()` method has already been implemented

File: tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    colossalai.launch(config=(dict()), rank=rank, world_size=world_size, port=port, host='localhost')
+    check_torch_ddp_checkpointIO()
+
+
+@rerun_if_address_is_in_use()
+def test_torch_ddp_checkpointIO():
+    spawn(run_dist, 2)
+
+
+def recursive_check(d1, d2):
```
Comment (ver217) [inline tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py:66]: This function can be extracted to a `utils` file

File: tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    for k, v in d1.items():
+        if isinstance(v, dict):
+            recursive_check(v, d2[k])
+        elif isinstance(v, list):
+            for i in range(len(v)):
+                if isinstance(v[i], torch.Tensor):
+                    v[i] = v[i].to('cpu')
+                    d2[k][i] = d2[k][i].to('cpu')
+                    if not torch.equal(v[i], d2[k][i]):
```
Comment (ver217) [inline tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py:75]: This can be simplified as `assert torch.equal(v[i], d2[k][i])`

File: tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                        raise AssertionError
+                elif v[i] != d2[k][i]:
+                    assert v[i] == d2[k][i]
+
+        elif isinstance(v, torch.Tensor):
+            v = v.to('cpu')
+            d2[k] = d2[k].to('cpu')
+            if not torch.equal(v, d2[k]):
```
Comment (ver217) [inline tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py:83]: Same as the above comment

File: colossalai/testing/comparison.py [hunk 5]
```diff
@@ -28,3 +28,23 @@ def assert_equal_in_group(tensor: Tensor, process_group: ProcessGroup = None):
         a = tensor_list[i]
         b = tensor_list[i + 1]
         assert torch.all(a == b), f'expected tensors on rank {i} and {i + 1} to be equal but they are not, {a} vs {b}'
+
+
+def recursive_check(d1, d2):
```
Comment (ver217) [inline colossalai/testing/comparison.py:33]: Please use a clearer name.

Comment (ver217) [reply colossalai/testing/comparison.py:33]: Like `check_state_dict_equal(d1: OrderedDict, d2: OrderedDict, ignore_device: bool = True)`

**Review by ver217 — APPROVED** [review]
